exercises/dnd-character/src/test/java/DnDCharacterTest.java

```diff
@@ -128,7 +133,224 @@ public void testRandomCharacterIsValid() {
+         * on choosing the actual values.
+         *
+         * Test setup just barely passes with:
+         * cycles:      relativeError:  constantError:
+         * 10000        0.00            0.016
+         * 10000        0.01            0.016
+         * 10000        0.02            0.013
+         * 10000        0.03            0.012
+         * 10000        0.04            0.009
+         * 10000        0.05            0.008
+         * 10000        0.06            0.007
+         * 10000        0.07            0.006
+         * 10000        0.08            0.006
+         * 10000        0.09            0.006
+         * 10000        0.10            0.006
+         *
+         * For a set of parameters that pass the test setup, increasing any one
+         * of the three, should still pass the test setup.
+         *
+         * The parameters (10000, 0.1, 0.01) were chosen based on the results
+         * of the table above, and were tested further in the test setup with a
+         * loop of 1000000, which passed.
+         *
+         * Further following should hold true:
+         *     relativeError <= 0.2
+         *     constantError <= 0.02
+         * Otherwise it becomes increasingly likely that an 'ability' method
+         * which gives "Throw three dice and sum them" would wrongly pass the
+         * test.
+         */
+        testDistribution(10000, 0.1, 0.01);
     }
 
+    private void testDistribution(int cycles, double relativeError, double constantError) {
+        if (cycles < 0 || relativeError < 0.0 || constantError < 0.0) {
+            throw new IllegalArgumentException();
+        }
+
+        int[] distribution = new int[19];
+        for (int i = 0; i < cycles; i++) {
+            int ability = dndCharacter.ability();
+            if (inRange(ability)) {
+                distribution[ability]++;
+            } else {
+                fail("ability out of Range");
+            }
+        }
+
+        for (int ability = 3; ability <= 18; ability++) {
+            double probability = (double) distribution[ability] / cycles;
+            double expected    = reference[ability];
+            double error       = Math.abs(probability - expected);
+            double errorBound  = relativeError * expected + constantError;
+            if (error > errorBound) {
+                fail();
+            }
+        }
+    }
+
+    private static final double[] reference;
+    static {
+        /* Amount of possible combinations for throwing 4 dice
+         */
+        int referenceCount = 6 * 6 * 6 * 6;
+        /* This distribution was generated by going through all possible
+         * combinations for throwing four dice, computing the ability for that
+         * combination, and incrementing the array at that index.
+         */
+        int[] referenceDist = new int[] {
+            0, 0, 0, 1, 4, 10, 21, 38, 62, 91, 122, 148, 167, 172, 160, 131, 94, 54, 21
+        };
```

Comment:
Instead use actual generation code? Or maybe put the code in a comment?

Reply:
I'm debating the merits. I assume the generation is time-prohibitive to just generate it in the test. We could leave it in an unused method just for documentation purposes?

Reply:
Is this comment bad?

If you ask me the ingenious part of this PR is that @jippiee precomputes the heavy parts and remains confident to any chosen probability that there isn't false negatives. But this array is certainly magical and deserves an explanation.

Reply:
Also, while some comments can be moved to the hint file, this one is overly specific about some magical constants. So I might be inclined to keep it local, since you really don't care about what the numbers are until you read them, but they also really need an explanation.

Reply:
That's not a bad idea, but not an entirely uncontentious one either.

Dead code vs. looking up commits for explanations.

I would be happy with either.

Reply:
I'm not saying we should go with the following, but it made me think of [golden tests](https://ro-che.info/articles/2017-12-04-golden-tests), i.e. unit tests stored in files.